sqlglot/dialects/starrocks.py

```diff
@@ -31,6 +78,68 @@ class Parser(MySQL.Parser):
             "REGEXP": exp.RegexpLike.from_arg_list,
         }
 
+        PROPERTY_PARSERS = {
+            **MySQL.Parser.PROPERTY_PARSERS,
+            "DISTRIBUTED BY HASH": lambda self: self._parse_distributed_by("HASH"),
+            "DISTRIBUTED BY RANDOM": lambda self: self._parse_distributed_by("RANDOM"),
+            "PROPERTIES": lambda self: self._parse_wrapped_csv(self._parse_property),
+            "DUPLICATE KEY": lambda self: self._parse_duplicate(),
+            "PARTITION BY": lambda self: self._parse_partitioned_by(),
+        }
+
+        def _parse_create(self) -> exp.Create | exp.Command:
+            create = super()._parse_create()
+
+            # Starrocks's primary is defined outside the schema, so we need to move it into the schema
+            # https://docs.starrocks.io/docs/table_design/table_types/primary_key_table/#usage
```

Comment:
Why do we need to move it inside the schema? What happens if we don't?

Reply:
If we do not change the outside primary key pattern, when we convert to hive (PRIMARY KEY starting from 2.1.0), hive only supports PRIMARY KEY specified inside the schema.
see: https://issues.apache.org/jira/browse/HIVE-13290

Reply:
That sounds like something that should be fixed in Hive's generator, not @ parse time for starrocks. Make sure to also test spark, databricks.

Reply:
Spark SQL does not support the setting of primary key. I used Spark 3.3 to test:
```
Exception in thread "main" org.apache.spark.sql.catalyst.parser.ParseException: 
Syntax error at or near 'PRIMARY'(line 2, pos 2)

== SQL ==
CREATE TABLE data_integration.foo2 (col1 int
) PRIMARY KEY (`col1`)
--^^^
;

	at org.apache.spark.sql.catalyst.parser.ParseException.withCommand(ParseDriver.scala:306)
	at org.apache.spark.sql.catalyst.parser.AbstractSqlParser.parse(ParseDriver.scala:143)
	at org.apache.spark.sql.execution.SparkSqlParser.parse(SparkSqlParser.scala:52)
	at org.apache.spark.sql.catalyst.parser.AbstractSqlParser.parsePlan(ParseDriver.scala:89)
	at org.apache.spark.sql.SparkSession.$anonfun$sql$2(SparkSession.scala:620)
	at org.apache.spark.sql.catalyst.QueryPlanningTracker.measurePhase(QueryPlanningTracker.scala:111)
	at org.apache.spark.sql.SparkSession.$anonfun$sql$1(SparkSession.scala:620)
	at org.apache.spark.sql.SparkSession.withActive(SparkSession.scala:779)
	at org.apache.spark.sql.SparkSession.sql(SparkSession.scala:617)
	at org.apache.spark.sql.SQLContext.sql(SQLContext.scala:651)
	at sparkDemo.sql.SQLDemo$.main(SQLDemo.scala:35)
	at sparkDemo.sql.SQLDemo.main(SQLDemo.scala)
```

Reply:
@georgesittas Oh, I understand what you mean. If this is a feature of the syntax itself, then it should be adapted in other generators instead of being fixed as a syntax bug, right?

But it seems that this pattern of specifying the primary key outside is not a universal standard except for doris and starrocks. 🤔

Reply:
For example, OLTP such as MySQL and PostgreSQL are all set up inside.

Reply:
If we are sure that this approach is not reasonable, we can delete it. I think I can submit a separate PR to fit it later =)

Reply:
Yes, I also checked mysql and postgres. It also supports the inside pattern:

```
mysql> CREATE TABLE foo1 (col1 int, PRIMARY KEY (`col1`)) ;
Query OK, 0 rows affected (0.03 sec)

mysql> CREATE TABLE foo1 (col1 int) PRIMARY KEY (`col1`);
ERROR 1064 (42000): You have an error in your SQL syntax; check the manual that corresponds to your MySQL server version for the right syntax to use near 'PRIMARY KEY (`col1`)' at line 1
```

Then, we can see this in the table_constraint of the postgresql document:
https://www.postgresql.org/docs/7.2/sql-createtable.html